confgenerator/confgenerator.go

```diff
@@ -97,6 +99,152 @@ func (m *Metrics) generateOtelPipelines() (map[string]otel.Pipeline, error) {
 	return out, nil
 }
 
+func (uc *UnifiedConfig) GenerateFluentBitConfigsModular(logsDir string, stateDir string, hostInfo *host.InfoStat) (string, string, error) {
+	userAgent, _ := getUserAgent("Google-Cloud-Ops-Agent-Logging", hostInfo)
+	components, err := uc.Logging.generateFluentbitComponents(userAgent, hostInfo)
+	if err != nil {
+		return "", "", err
+	}
+
+	c := fluentbit.ModularConfig{
+		Variables: map[string]string{
+			"buffers_dir": path.Join(stateDir, "buffers"),
+			"logs_dir":    logsDir,
+		},
+		Components: components,
+	}
+	return c.Generate()
+}
+
+type fbSource struct {
+	tag        string
+	components []fluentbit.Component
+}
+
+func (l *Logging) generateFluentbitComponents(userAgent string, hostInfo *host.InfoStat) ([]fluentbit.Component, error) {
+	var out []fluentbit.Component
+	out = append(out, fluentbit.Service{}.Component())
+
+	if l != nil && l.Service != nil {
+		var sources []fbSource
+		var logNames []string
+		for pID, p := range l.Service.Pipelines {
+			for _, rID := range p.ReceiverIDs {
+				receiver, ok := l.Receivers[rID]
+				if !ok {
+					return nil, fmt.Errorf("receiver %q not found", rID)
+				}
+				tag := fmt.Sprintf("%s.%s", pID, rID)
+				components := receiver.Components(tag)
+				for i, pID := range p.ProcessorIDs {
+					processor, ok := l.Processors[pID]
+					if !ok {
+						processor, ok = LegacyBuiltinProcessors[pID]
+					}
+					if !ok {
+						return nil, fmt.Errorf("processor %q not found", pID)
+					}
+					components = append(components, processor.Components(tag, i)...)
+				}
+				components = append(components, setLogNameComponents(tag, rID)...)
+				logNames = append(logNames, regexp.QuoteMeta(rID))
+				sources = append(sources, fbSource{tag, components})
+			}
+		}
+		sort.Slice(sources, func(i, j int) bool { return sources[i].tag < sources[j].tag })
+		sort.Strings(logNames)
+
+		for _, s := range sources {
+			out = append(out, s.components...)
+		}
+		if len(logNames) > 0 {
+			out = append(out, fluentbit.Stackdriver{
+				Match:     strings.Join(logNames, "|"),
+				Workers:   8,
+				UserAgent: userAgent,
+			}.Component())
+		}
+	}
+	// TODO: Use receivers instead of generating Tail objects directly.
+	if hostInfo.OS != "windows" {
+		out = append(out, fluentbit.Tail{
+			Tag:          "ops-agent-collectd",
+			IncludePaths: []string{"${logs_dir}/metrics-module.log"},
+		}.Component())
+	}
+	out = append(out, fluentbit.Tail{
+		Tag:          "ops-agent-fluent-bit",
+		IncludePaths: []string{"${logs_dir}/logging-module.log"},
+	}.Component())
+	out = append(out, fluentbit.Stackdriver{
+		Match:     "ops-agent-fluent-bit|ops-agent-collectd",
+		Workers:   8,
+		UserAgent: userAgent,
+	}.Component())
+
+	return out, nil
+}
+
+func setLogNameComponents(tag, logName string) []fluentbit.Component {
+	// TODO: Can we just set log_name_key in the output plugin and avoid this mess?
```

Comment:
If we consistently append the pipeline ID in the tag, we can strip off anything before the first `.` in the output plugin. That should work as well.

Alternatively, something like below might also work (not verified though):

Current config
```
[FILTER]
    Add   logName log_source_id1
    Match pipeline1.log_source_id1
    Name  modify

[FILTER]
    Emitter_Mem_Buf_Limit 10M
    Emitter_Storage.type  filesystem
    Match                 pipeline1.log_source_id1
    Name                  rewrite_tag
    Rule                  $logName .* $logName false

[FILTER]
    Match  log_source_id1
    Name   modify
    Remove logName
```

Alternative config
https://docs.fluentbit.io/manual/pipeline/filters/rewrite-tag#regex
// Match pipeline_id as group 1 and receiver_id as group 2 and rewrite the tag to group 2.
```
[FILTER]
    Emitter_Mem_Buf_Limit 10M
    Emitter_Storage.type  filesystem
    Match                 pipeline1.log_source_id1
    Name                  rewrite_tag
    Rule                  $logName ^([^.]*)\.(.*)$ $2 false
```

I'm leaning towards the former though for simplicity.

Reply:
I'm pretty sure fluentbit would let us set `log_name_key logName` in the output plugin and it would just consume the `logName` label directly. I prefer that over anything involving regexes.

Reply:
If we can indeed do this, it would be a very nice cleanup, but it's probably best to defer it to a follow-up PR. Also, we've been lucky so far that nobody complained about us tromping all over the `logName` field — there's a reason the default `log_name_key` value is `logging.googleapis.com/logName` and not just `logName`.

Reply:
Hence why I left it as a `TODO`.